ocaml/xapi/storage_utils.ml

```diff
@@ -17,3 +17,36 @@ let string_of_vdi_type vdi_type =
 
 let vdi_type_of_string str =
   API.vdi_type_of_rpc (Rpc.String str)
+
+module D=Debug.Make(struct let name="storage_utils" end)
+open D
+
+(** [redirectable_rpc ~srcstr ~dststr ~remote_url_of_ip ~local_fn call] is an RPC function
+    that first attempts to call [local_fn call], and if that returns a [Redirect ip] exception then
+    it sends a remote RPC to the url built by [remote_url_of_ip].
+ *)
+let redirectable_rpc ~srcstr ~dststr ~remote_url_of_ip ~local_fn =
+  let rec rpc ~f call =
+    (* on first iteration this will be the [local_fn] supplied by the caller *)
+    let result = f call in
+    if result.Rpc.success then result
+    else begin
+        debug "Got failure: checking for redirect";
```

Comment:
This spreads log messages that cannot be identified individually over several log lines which might no longer be next to each other in the log. Log should be self-contained.

Reply:
Agreed, I'll fix this